File: task5/main.py
# Вариант с файлом:
# with open("input.txt") as f:
#     lines = f.read().splitlines()
#     n, m = map(int, lines[0].split())
#     graph = {i: [] for i in range(n)}
#     for line in lines[1:m+1]:
#         u, v = map(int, line.split())
#         graph[u].append(v)

def find_universal_source(graph, n):
    def dfs(start):
        visited = set()
        stack = [start]
        while stack:
            node = stack.pop()
            if node not in visited:
                visited.add(node)
                for neighbor in graph[node]:
                    if neighbor not in visited:
                        stack.append(neighbor)
        return visited

    for v in range(n):
        reachable = dfs(v)
        if len(reachable) == n:
            return v

    return -1

n, m = map(int, input("Введите количество вершин и рёбер: ").split())
graph = {i: [] for i in range(n)}
for _ in range(m):
    u, v = map(int, input("Введите ребро (u v): ").split())
    graph[u].append(v)

result = find_universal_source(graph, n)
if result != -1:
    print(f"Вершина {result} достигает все остальные")
else:
    print("Такой вершины нет")

File: task5/main.cpp
#include <iostream>
#include <vector>
#include <stack>
// Вариант с файлом:
// #include <fstream>
using namespace std;

int countReachable(vector<vector<int>>& graph, int start, int n) {
    vector<bool> visited(n, false);
    stack<int> st;
    st.push(start);

    while (!st.empty()) {
        int node = st.top();
        st.pop();
        if (!visited[node]) {
            visited[node] = true;
            for (int neighbor : graph[node]) {
                if (!visited[neighbor]) {
                    st.push(neighbor);
                }
            }
        }
    }

    int count = 0;
    for (bool v : visited) if (v) count++;
    return count;
}

int main() {
    int n, m;
    cout << "Введите количество вершин и рёбер: ";
    cin >> n >> m;

    // Вариант с файлом:
    // ifstream f("input.txt");
    // f >> n >> m;
    // vector<vector<int>> graph(n);
    // for (int i = 0; i < m; i++) {
    //     int u, v; f >> u >> v;
    //     graph[u].push_back(v);
    // }

    vector<vector<int>> graph(n);
    for (int i = 0; i < m; i++) {
        int u, v;
        cout << "Введите ребро (u v): ";
        cin >> u >> v;
        graph[u].push_back(v);
    }

    for (int v = 0; v < n; v++) {
        if (countReachable(graph, v, n) == n) {
            cout << "Вершина " << v << " достигает все остальные" << endl;
            return 0;
        }
    }

    cout << "Такой вершины нет" << endl;
    return 0;
}